asset-transfer-private-data/application-gateway-go/app.go

```diff
@@ -0,0 +1,358 @@
+}
+
+func agreeToTransfer(contract client.Contract, assetID string) {
+	// Buyer from Org2 agrees to buy the asset.
+	// To purchase the asset, the buyer needs to agree to the same value as the asset owner.
+	dataForAgreement := struct {
+		AssetID        string
+		AppraisedValue int
+	}{assetID, 100}
+	fmt.Printf("\n--> Submit Transaction: AgreeToTransfer, payload: %+v\n", dataForAgreement)
+
+	if _, err := contract.Submit(
+		"AgreeToTransfer",
+		client.WithTransient(transientData("asset_value", dataForAgreement)),
+	); err != nil {
+		panic(err)
+	}
+
+	logTxCommitSuccess()
+}
+
+func readTransferAgreement(contract client.Contract, assetID string) {
+	fmt.Printf("\n--> Evaluate Transaction: ReadTransferAgreement, ID: %s\n", assetID)
+
+	resultBytes, err := contract.EvaluateTransaction("ReadTransferAgreement", assetID)
+	if err != nil {
+		panic(err)
+	}
+
+	result := formatJSON(resultBytes)
+	if result == "" {
+		doFail("Received empty result for ReadTransferAgreement")
+	}
+	fmt.Printf("*** Result: %s\n", result)
+}
+
+func transferAsset(contract client.Contract, assetID string) (err error) {
+	fmt.Printf("\n--> Submit Transaction: TransferAsset, ID: %s\n", assetID)
+
+	buyerDetails := struct {
+		AssetID  string
+		BuyerMSP string
+	}{assetID, mspIDOrg2}
+
+	if _, err = contract.Submit(
+		"TransferAsset",
+		client.WithTransient(
+			transientData("asset_owner", buyerDetails),
+		),
+	); err != nil {
+		return
+	}
+
+	logTxCommitSuccess()
+	return
+}
+
+func deleteAsset(contract client.Contract, assetID string) (err error) {
+	fmt.Printf("\n--> Submit Transaction: DeleteAsset, ID: %s\n", assetID)
+
+	dataForDelete := struct{ AssetID string }{assetID}
+
+	if _, err = contract.Submit(
+		"DeleteAsset",
+		client.WithTransient(
+			transientData("asset_delete", dataForDelete),
+		),
+	); err != nil {
+		return
+	}
+
+	logTxCommitSuccess()
+	return
+}
+
+func purgeAsset(contract client.Contract, assetID string) (err error) {
+	fmt.Printf("\n--> Submit Transaction: PurgeAsset, ID: %s\n", assetID)
+
+	dataForPurge := struct{ AssetID string }{assetID}
+
+	if _, err = contract.Submit(
+		"PurgeAsset",
+		client.WithTransient(
+			transientData("asset_purge", dataForPurge),
+		),
+	); err != nil {
+		return
+	}
+
+	logTxCommitSuccess()
+	return
+}
+
+func readAssetPrivateDetails(contract client.Contract, assetID, collectionName string) bool {
+	fmt.Printf("\n--> Evaluate Transaction: ReadAssetPrivateDetails from %s, ID: %s\n", collectionName, assetID)
+
+	resultBytes, err := contract.EvaluateTransaction("ReadAssetPrivateDetails", collectionName, assetID)
+	if err != nil {
+		panic(err)
+	}
+
+	result := formatJSON(resultBytes)
+	if result == "" {
+		fmt.Println("*** No result")
+		return false
+	}
+	fmt.Printf("*** Result: %s\n", result)
+	return true
+}
+
+func transientData(key string, value any) map[string][]byte {
+	valueAsBytes, err := json.Marshal(&value)
+	if err != nil {
+		panic(err)
+	}
+
+	return map[string][]byte{key: valueAsBytes}
+}
```

Comment:
This definitely does the job for this use-case, where the application only creates transient maps with a single key/value. I wonder if (as an example) it might be worth using a more flexible alternative, something like:

```go
func asTransientJSON(data map[string]any) map[string][]byte {
	results := make(map[string][]byte)

	for key, value := range data {
		if jsonValue, err := json.Marshal(value); err != nil {
			panic(err)
		} else {
			results[key] = jsonValue
		}
	}

	return results
}
```

This would be used as follows:

```go
client.WithTransient(asTransientJSON(map[string]any{
	"asset_properties": asset1Data,
}))
```

Reply:
Yes, good point and good idea. I would add a type alias to the map so it looks a bit nicer:

```go
type transient = map[string][]byte

// ...
client.WithTransient(transient{
	"asset_properties": marshal(asset1Data),
}),
```
wdyt?